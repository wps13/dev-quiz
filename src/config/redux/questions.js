import questions from '../../utils/questions';
import calculateResult from '../../utils/calculateResult';

const ADD_ANSWER = 'ADD_ANSWER';
const RESET_DATA = 'RESET_DATA';

export const changeQuestion = answerText => ({
  type: ADD_ANSWER,
  payload: {
    answerText,
  },
});

export const reset = () => ({type: RESET_DATA});

const INITIAL_STATE = {
  allQuestions: questions,
  currentQuestion: 0,
  answers: [],
  actualQuestion: questions[0],
  actualQuestionId: 1,
  lastQuestion: false,
  totalQuestions: questions.length,
  correctAnswers: 0,
  minimumPercent: 75,
  correctPercent: 0,
  remainingQuestions: questions.length - 1,
  status: '',
};

const questionReducer = (action, state = INITIAL_STATE) => {
  switch (action.type) {
    case ADD_ANSWER:
      let newAnswers = state.answers.concat({
        id: state.actualQuestionId,
        answer: action.payload.answerText,
      });
      let currentQuestion = state.currentQuestion;
      let totalQ = state.totalQuestions;
      let minimumPerc = state.minimumPercent;
      const ques = state.allQuestions;
      let [corrects, percentCorrect, currentStatus] = calculateResult({
        questions: ques,
        answers: newAnswers,
        totalQuestions: totalQ,
        minimumPercent: minimumPerc,
      });
      let remainingQ = state.remainingQuestions;
      let actualQID = state.actualQuestionId;

      return {
        ...state,
        currentQuestion: currentQuestion + 1,
        answers: newAnswers,
        actualQuestionId: actualQID + 1,
        actualQuestion: questions[currentQuestion + 1],
        correctAnswers: corrects,
        correctPercent: percentCorrect,
        remainingQuestions: remainingQ - 1,
        status: currentStatus,
        lastQuestion: actualQID + 1 >= totalQ,
      };
    case RESET_DATA:
      return {
        ...state,
        answers: [],
        actualQuestion: state.allQuestions[0],
        actualQuestionId: 1,
        lastQuestion: false,
        currentQuestion: 0,
        correctAnswers: 0,
        correctPercent: 0,
        status: '',
        remainingQuestions: state.allQuestions.length - 1,
      };
    default:
      return state;
  }
};

export default questionReducer;
