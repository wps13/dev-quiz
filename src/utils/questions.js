export default [
  {
    id: 1,
    title: 'Which tag is used to insert an image in html?',
    alternatives: ['p', 'img', 'div', 'media'],
    answer: 'img',
  },
  {
    id: 2,
    title: 'How does position absolute works on css?',
    alternatives: [
      'It moves the element out of the document flow',
      'it moves the elements to another parent',
      'it hides the elements',
      'it puts the element above all the other elements',
    ],
    answer: 'It moves the element out of the document flow',
  },
  {
    id: 3,
    title: 'How to grow the flex container?',
    alternatives: ['flex-grow', 'flex-flow', 'flex', 'flex-wrap'],
    answer: 'flex-grow',
  },
];
