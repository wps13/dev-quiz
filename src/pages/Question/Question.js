import React, {useState} from 'react';
import {View, Text, StyleSheet, FlatList, SafeAreaView} from 'react-native';
import {useSelector, useDispatch} from 'react-redux';

import Button from '../../components/Button/Button';
import COLORS from '../../utils/COLORS';
import {baseStyles} from '../../assets/styles';
import {changeQuestion} from '../../config/redux/questions';

const Question = ({navigation}) => {
  const dispatch = useDispatch();
  const questions = useSelector(state => state.question);
  const [shouldRender, setShouldRender] = useState(true);

  const goToResult = () => {
    if (questions.lastQuestion) {
      setShouldRender(false);
      return navigation.navigate('Results');
    }
    return null;
  };

  const handleQuestionChange = text => {
    dispatch(changeQuestion(text));
    goToResult();
  };

  const renderAlternative = ({index, item}) => (
    <Button
      textColor={COLORS.white}
      buttonColor={COLORS.darkPurple}
      text={item}
      key={`${item}${index}`}
      pressFunction={handleQuestionChange}
    />
  );
  return (
    <>
      {shouldRender && (
        <SafeAreaView style={styles.container}>
          <Text style={styles.title}>
            Question {questions.actualQuestionId}
          </Text>
          <Text style={styles.text}>{questions.actualQuestion.title}</Text>
          <View style={styles.container}>
            <FlatList
              data={questions.actualQuestion.alternatives}
              renderItem={renderAlternative}
              keyExtractor={item => item}
              extraData={questions}
            />
          </View>
          <View>
            <Text>Remaining questions: {questions.remainingQuestions} </Text>
          </View>
        </SafeAreaView>
      )}
    </>
  );
};
export default Question;

const styles = StyleSheet.create({
  container: baseStyles.container,
  title: baseStyles.title,
  text: {
    fontSize: 16,
  },
});
