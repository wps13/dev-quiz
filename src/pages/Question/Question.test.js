import React from 'react';
import {shallow, mount} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import Question from './Question';

import questions from '../../../__mocks__/questions';

const defaultValue = {
  allQuestions: questions,
  currentQuestion: 0,
  answers: [],
  actualQuestion: questions[0],
  actualQuestionId: 1,
  lastQuestion: false,
  totalQuestions: questions.length,
  correctAnswers: 0,
  minimumPercent: 75,
  correctPercent: 0,
  remainingQuestions: 1,
  status: '',
};

const mockStore = configureMockStore([]);

const props = {
  navigation: {
    navigate: jest.fn(),
  },
};

let store;

const renderQuestion = () =>
  shallow(
    <Provider store={store}>
      <Question {...props} />
    </Provider>,
  );

describe('Question', () => {
  beforeEach(() => {
    store = mockStore({question: defaultValue});
    store.dispatch = jest.fn();
  });
  it('should render correctly', () => {
    const wrapper = renderQuestion();

    expect(wrapper).toMatchSnapshot();
  });
  it('should render correctly the remaining questions', () => {
    jest.mock('react-redux', () => ({
      useSelector: selectorFn =>
        selectorFn({
          question: {
            allQuestions: [
              {
                id: 1,
                title: 'Which tag is used to insert an image in html?',
                alternatives: ['p', 'img', 'div', 'media'],
                answer: 'img',
              },
              {
                id: 2,
                title: 'How does position absolute works on css?',
                alternatives: [
                  'It moves the element out of the document flow',
                  'it moves the elements to another parent',
                  'it hides the elements',
                  'it puts the element above all the other elements',
                ],
                answer: 'It moves the element out of the document flow',
              },
            ],
            currentQuestion: 0,
            answers: [],
            actualQuestion: {
              id: 1,
              title: 'Which tag is used to insert an image in html?',
              alternatives: ['p', 'img', 'div', 'media'],
              answer: 'img',
            },
            actualQuestionId: 1,
            lastQuestion: false,
            totalQuestions: 2,
            correctAnswers: 0,
            minimumPercent: 75,
            correctPercent: 0,
            remainingQuestions: 1,
            status: '',
          },
        }),
      useDispatch: jest.fn(),
    }));

    const wrapper = mount(
      <Provider store={store}>
        <Question {...props} />
      </Provider>,
    );

    const remainingText = wrapper.findWhere(
      node => node.prop('testID') === 'Question-RemainingValue',
    );

    expect(remainingText.exists()).toEqual(true);
    expect(remainingText.first().text()).toEqual('Remaining questions: 1');
  });
});
