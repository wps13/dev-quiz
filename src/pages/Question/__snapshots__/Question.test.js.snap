// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Question
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
  />
</ContextProvider>
`;
